examples/model-viewer/main.cpp

```diff
@@ -0,0 +1,922 @@
+            }
+
+            auto& lightTypeArray = lightArrays[tt];
+            size_t lightCount = lightTypeArray->lights.size();
+            size_t maxLightCount = lightTypeArray->layout.maximumCount;
+
+            // Recall that we are representing the data for a single
+            // light type `L` as either an instance of type `L` (if
+            // only a single light is supported), or as an instance
+            // of the type `LightArray<L,N>`.
+            //
+            if (maxLightCount == 1)
+            {
+                // This is the case where the maximu number of lights of
+                // the given type was set as one, so we just have a value
+                // of type `L`, and can tell the first light in our application-side
+                // array to encode itself into that location.
+
+                if (lightCount > 0)
+                {
+                    lightTypeCursor.setObject(
+                        lightTypeArray->lights[0]->createShaderObject(context));
+                }
+                else
+                {
+                    // We really ought to zero out the entry in this case
+                    // (under the assumption that all zeros will represent
+                    // an inactive light).
+                }
+            }
+            else
+            {
+                // The more interesting case is when we have a `LightArray<L,N>`,
+                // in which case we need to fill in the first field (the light count)...
+                //
+                uint32_t lightCount = uint32_t(lightTypeArray->lights.size());
+                lightTypeCursor["count"].setData(&lightCount, sizeof(lightCount));
+                //
+                // ... followed by an array of values of type `L` in the second field.
+                // We will only write to the first `lightCount` entries, which may be
+                // less than `N`. We will rely on dynamic looping in the shader to
+                // not access the entries past that point.
+                //
+                auto arrayCursor = lightTypeCursor["lights"];
+                for (size_t ii = 0; ii < lightCount; ++ii)
+                {
+                    arrayCursor[ii].setObject(
+                        lightTypeArray->lights[ii]->createShaderObject(context));
+                }
+            }
+        }
+        return shaderObject;
+    }
+};
+
+// Now that we've written all the required infrastructure code for
+// the application's renderer and shader library, we can move on
+// to the main logic.
+//
+// We will again structure our example application as a C++ `struct`,
+// so that we can scope its allocations for easy cleanup, rather than
+// use global variables.
+//
+struct ModelViewer : WindowedAppBase
+{
+
+RendererContext context;
+
+// Most of the application state is stored in the list of loaded models,
+// as well as the active light source (a single light for now).
```

Comment:
This comment is incorrect when it refers to a "single light". The comments in general will need a pass before we treat this example as being enabled/published again.